diff --git a/filemanager/UploadHandler.php b/filemanager/UploadHandler.php
index a45ea14..ff79f68 100755
--- a/filemanager/UploadHandler.php
+++ b/filemanager/UploadHandler.php
@@ -1144,6 +1144,13 @@ class UploadHandler
                 } else {
                     move_uploaded_file($uploaded_file, $file_path);
                 }
+            }elseif(file_exists($uploaded_file)){ // Если загруженный файл есть
+                if (empty($this->options['ftp'])) // и мы не используем ftp
+                    rename($uploaded_file, $file_path); // переименовываем файл
+                else{// если используем ftp -  нужно забросиь файл на сервер
+                    $fn=$this->options['storeFolder'].fix_filename(basename($file_path),$this->options['config']);
+                    $this->options['ftp']->put($fn,$uploaded_file,FTP_BINARY);
+                }
             } else {
                 // Non-multipart uploads (PUT method support)
                 file_put_contents(
@@ -1389,6 +1396,9 @@ class UploadHandler
         $content_range = $content_range_header ?
             preg_split('/[^0-9]+/', $content_range_header) : null;
         $size =  $content_range ? $content_range[3] : null;
+        // Не грузились файлы мелкого размера
+        if (!isset($content_range))
+            $content_range=[0,$size,$size];
         $files = array();
         if ($upload) {
             if (is_array($upload['tmp_name'])) {
diff --git a/filemanager/ajax_calls.php b/filemanager/ajax_calls.php
index e514186..7a7efb9 100755
--- a/filemanager/ajax_calls.php
+++ b/filemanager/ajax_calls.php
@@ -1,7 +1,5 @@
 <?php
 
-$config = include 'config/config.php';
-
 require_once 'include/utils.php';
 
 if ($_SESSION['RF']["verify"] != "RESPONSIVEfilemanager") {
diff --git a/filemanager/dialog.php b/filemanager/dialog.php
index 3ebde34..7b0db0d 100644
--- a/filemanager/dialog.php
+++ b/filemanager/dialog.php
@@ -1,8 +1,6 @@
 <?php
 $time = time();
 
-$config = include 'config/config.php';
-
 if (USE_ACCESS_KEYS == true){
 	if (!isset($_GET['akey'], $config['access_keys']) || empty($config['access_keys'])){
 		die('Access Denied!');
@@ -597,9 +595,15 @@ $src = '';
 if ($ftp) {
     try {
         $files = $ftp->scanDir($config['ftp_base_folder'] . $config['upload_dir'] . $rfm_subfolder . $subdir);
-        if (!$ftp->isDir($config['ftp_base_folder'] . $config['ftp_thumbs_dir'] . $rfm_subfolder . $subdir)) {
+        if (!$ftp->isDir($config['ftp_base_folder'] . $config['ftp_thumbs_dir'] . $rfm_subfolder . $subdir))
             create_folder(false, $config['ftp_base_folder'] . $config['ftp_thumbs_dir'] . $rfm_subfolder . $subdir, $ftp, $config);
-        }
+
+        // сканируем наличие тумбсов ...
+        $filesThumbs=[];
+        foreach($ftp->scanDir($config['ftp_base_folder'] . $config['ftp_thumbs_dir'] . $rfm_subfolder . $subdir) as $f)
+            if ($f['type']=='file')
+                $filesThumbs[]=$f['name'];
+
     } catch (FtpClient\FtpException $e) {
         echo "Error: ";
         echo $e->getMessage();
@@ -1073,6 +1077,15 @@ $files = $sorted;
 
                     $img_width = $img_height = "";
                     if($ftp){
+                        // если под файл $file не нашелся тумбс пробуем создать  такой файл .. .
+                        if (!in_array($file, $filesThumbs)){
+                            $creation_thumb_path = '/'.$config['ftp_base_folder'].$config['ftp_thumbs_dir'].$subdir. $file;
+                            // победили глюк с пробелами в назнвании файлов ..
+                            $bn=basename($src);
+                            $src=str_replace($bn, rawurlencode($bn), $src);
+                            if (!create_img($src, $creation_thumb_path, 122, 91, 'crop', $config,$ftp))
+                                $src_thumb = $mini_src = "";
+                        }
                         $mini_src = $src_thumb = $config['ftp_base_url'].$config['ftp_thumbs_dir'].$subdir. $file;
                         $creation_thumb_path = "/".$config['ftp_base_folder'].$config['ftp_thumbs_dir'].$subdir. $file;
                     }else{
@@ -1097,6 +1110,12 @@ $files = $sorted;
                         }
                     }
                 }
+                // зачищаем пути от document root ... чтобы  все они были относительно корня ..
+                foreach(['src_thumb','src','mini_src']as $var){
+                    $$var=str_replace($_SERVER['DOCUMENT_ROOT'], '', $$var);
+                    if ($thumbsAsset && strpos($$var, $thumbsAsset->sourcePath)===0)
+                        $$var=str_replace($thumbsAsset->sourcePath,$thumbsAsset->baseUrl , $$var);
+                }
                 $is_icon_thumb=false;
                 $is_icon_thumb_mini=false;
                 $no_thumb=false;
diff --git a/filemanager/execute.php b/filemanager/execute.php
index 0321bfa..7e7b878 100755
--- a/filemanager/execute.php
+++ b/filemanager/execute.php
@@ -1,5 +1,4 @@
 <?php
-$config = include 'config/config.php';
 
 include 'include/utils.php';
 
@@ -101,6 +100,10 @@ if (isset($info['extension']) && !(isset($_GET['action']) && $_GET['action'] ==
     exit;
 }
 
+// при ипользовани ftp .. дорисовываем к пути базовый катаог ftp чтобы путь стал абсолютным ..
+if ($ftp && !empty($config['ftp_base_folder']) && $path)
+    $path=DIRECTORY_SEPARATOR.$config['ftp_base_folder'].$path;
+
 if (isset($_GET['action'])) {
     switch ($_GET['action']) {
         case 'delete_file':
@@ -151,14 +154,16 @@ if (isset($_GET['action'])) {
 				$path .= $name;
 				$path_thumb .= $name;
 				$res = create_folder(fix_path($path,$config),fix_path($path_thumb,$config),$ftp,$config);
-				if(!$res){
+                // не буем матюкаться если  использовем ftp ..
+				if(!$ftp && !$res){
 					response(trans('Rename_existing_folder').AddErrorLocation())->send();
 				}
 			}
 			break;
 		case 'rename_folder':
 			if ($config['rename_folders']){
-                if(!is_dir($path)) {
+                // не буем матюкаться если  использовем ftp ..
+                if(!$ftp && !is_dir($path)) {
                     response(trans('wrong path').AddErrorLocation())->send();
                     exit;
                 }
diff --git a/filemanager/force_download.php b/filemanager/force_download.php
index 42f1666..f63512e 100755
--- a/filemanager/force_download.php
+++ b/filemanager/force_download.php
@@ -1,6 +1,5 @@
 <?php
 
-$config = include 'config/config.php';
 
 include 'include/utils.php';
 include 'include/mime_type_lib.php';
diff --git a/filemanager/include/utils.php b/filemanager/include/utils.php
index 8cda960..4f1bebc 100755
--- a/filemanager/include/utils.php
+++ b/filemanager/include/utils.php
@@ -1,4 +1,6 @@
 <?php
+// для поддержки языка ...
+global $lang_vars;
 
 if (!isset($_SESSION['RF']) || $_SESSION['RF']["verify"] != "RESPONSIVEfilemanager")
 {
@@ -318,6 +320,8 @@ function rename_file($old_path, $name, $ftp = null, $config = null)
 
 function url_exists($url){
     if (!$fp = curl_init($url)) return false;
+    // закрываем лоткрытый curl чтобы не висел ...
+    curl_close($fp);
     return true;
 }
 
@@ -364,10 +368,6 @@ function rename_folder($old_path, $name, $ftp = null, $config = null)
 
 function ftp_con($config){
 	if(isset($config['ftp_host']) && $config['ftp_host']){
-		// *** Include the class
-		include('include/FtpClient.php');
-		include('include/FtpException.php');
-		include('include/FtpWrapper.php');
 
 		$ftp = new \FtpClient\FtpClient();
 		try{
@@ -401,10 +401,10 @@ function ftp_con($config){
 * @return bool
 * @throws \Exception
 */
-function create_img($imgfile, $imgthumb, $newwidth, $newheight = null, $option = "crop",$config = array())
+function create_img($imgfile, $imgthumb, $newwidth, $newheight = null, $option = "crop",$config = array(), $ftp=null )
 {
 	$result = false;
-	if(isset($config['ftp_host']) && $config['ftp_host']){
+	if($ftp){
 		if(url_exists($imgfile)){
 			$temp = tempnam('/tmp','RF');
 			unlink($temp);
@@ -565,7 +565,8 @@ function checkresultingsize($sizeAdded)
 function create_folder($path = null, $path_thumbs = null,$ftp = null,$config = null)
 {
 	if($ftp){
-		$ftp->mkdir($path);
+		// рекурсивно создаём каталоги ..
+		$ftp->mkdir($path,true);
 		$ftp->mkdir($path_thumbs);
 	}else{
 		if(file_exists($path) || file_exists($path_thumbs)){
@@ -877,20 +878,28 @@ function image_check_memory_usage($img, $max_breedte, $max_hoogte)
 			$memory_limit = 0;
 			if (strpos($mem, 'M') !== false) $memory_limit = abs(intval(str_replace(array('M'), '', $mem) * 1024 * 1024));
 			if (strpos($mem, 'G') !== false) $memory_limit = abs(intval(str_replace(array('G'), '', $mem) * 1024 * 1024 * 1024));
+			// для числового задания параметра memory_limit
+			if (preg_match('#^\d+$#',$mem))$memory_limit=intval($mem);
 			
-			$image_properties = getimagesize($img);
-			$image_width = $image_properties[0];
-			$image_height = $image_properties[1];
-			if (isset($image_properties['bits']))
-				$image_bits = $image_properties['bits'];
-			else
-				$image_bits = 0;
-			$image_memory_usage = $K64 + ($image_width * $image_height * ($image_bits >> 3) * 2);
-			$thumb_memory_usage = $K64 + ($max_breedte * $max_hoogte * ($image_bits >> 3) * 2);
-			$memory_needed = abs(intval($memory_usage + $image_memory_usage + $thumb_memory_usage));
+			// на всякий пожарный обернули в try ...
+			try{
+				$image_properties = getimagesize($img);
+				$image_width = $image_properties[0];
+				$image_height = $image_properties[1];
+				if (isset($image_properties['bits']))
+					$image_bits = $image_properties['bits'];
+				else
+					$image_bits = 0;
+				$image_memory_usage = $K64 + ($image_width * $image_height * ($image_bits >> 3) * 2);
+				$thumb_memory_usage = $K64 + ($max_breedte * $max_hoogte * ($image_bits >> 3) * 2);
+				$memory_needed = abs(intval($memory_usage + $image_memory_usage + $thumb_memory_usage));
 
-			if ($memory_needed > $memory_limit)
-			{
+				if ($memory_needed > $memory_limit)
+				{
+					return false;
+				}
+			}catch(\Exception $e){
+				\Yii::info($e->getMessage(),'err');
 				return false;
 			}
 		}
diff --git a/filemanager/upload.php b/filemanager/upload.php
index 639f51d..7f4ceb1 100755
--- a/filemanager/upload.php
+++ b/filemanager/upload.php
@@ -1,9 +1,5 @@
 <?php
-
 try {
-    if (!isset($config)) {
-        $config = include 'config/config.php';
-    }
 
     include 'include/utils.php';
 
@@ -88,7 +84,7 @@ try {
                 'name' => array(basename($_POST['url'])),
                 'tmp_name' => array($temp),
                 'size' => array(filesize($temp)),
-                'type' => null
+                'type' => mime_content_type($temp),// определились с mime типом ..
             );
         } else {
             throw new Exception('Is not a valid URL.');
@@ -141,7 +137,7 @@ try {
         'storeFolder' => $storeFolder,
         'storeFolderThumb' => $storeFolderThumb,
         'ftp' => $ftp,
-        'upload_dir' => dirname($_SERVER['SCRIPT_FILENAME']) . '/' . $storeFolder,
+        'upload_dir' => $storeFolder,
         'upload_url' => $config['base_url'] . $config['upload_dir'] . $_POST['fldr'],
         'mkdir_mode' => $config['folderPermission'],
         'max_file_size' => $config['MaxSizeUpload'] * 1024 * 1024,
